appinventor/components/src/com/google/appinventor/components/runtime/Navigation.java

```diff
@@ -0,0 +1,352 @@
+    } else if (!isValidLongitude(longitude)) {
+      getDispatchDelegate().dispatchErrorOccurredEvent(this, "SetStartLocation",
+          ErrorMessages.ERROR_INVALID_LONGITUDE, longitude);
+    } else {
+      startLocation.setCoords(latitude, longitude);
+    }
+  }
+
+  @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_LATITUDE,
+      defaultValue = "0.0")
+  @SimpleProperty
+  public void EndLatitude(double latitude) {
+    if (isValidLatitude(latitude)) {
+      endLocation.setLatitude(latitude);
+    } else {
+      getDispatchDelegate().dispatchErrorOccurredEvent(this, "EndLatitude",
+          ErrorMessages.ERROR_INVALID_LATITUDE, latitude);
+    }
+  }
+
+  @SimpleProperty(category = PropertyCategory.BEHAVIOR,
+      description = "The latitude of the end location.")
+  public double EndLatitude() {
+    return endLocation.getLatitude();
+  }
+
+  @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_LONGITUDE,
+      defaultValue = "0.0")
+  @SimpleProperty
+  public void EndLongitude(double longitude) {
+    if (isValidLongitude(longitude)) {
+      endLocation.setLongitude(longitude);
+    } else {
+      getDispatchDelegate().dispatchErrorOccurredEvent(this, "EndLongitude",
+          ErrorMessages.ERROR_INVALID_LONGITUDE, longitude);
+    }
+  }
+
+  @SimpleProperty(category = PropertyCategory.BEHAVIOR,
+      description = "The longitude of the end location.")
+  public double EndLongitude() {
+    return endLocation.getLongitude();
+  }
+
+  @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_NAVIGATION_METHOD,
+      defaultValue = "foot-walking")
+  @SimpleProperty
+  public void TransportationMethod(String method) {
+    for (TransportMethod t : TransportMethod.values()) {
+      if (method.equals(t.method())) {
+        this.method = t;
+      }
+    }
+  }
+
+  @SimpleProperty(category = PropertyCategory.BEHAVIOR,
+      description = "The longitude of the end location.")
+  public String TransportationMethod() {
+    return method.method();
+  }
+
+  @SimpleFunction(description = "Set the end location.")
+  public void SetEndLocation(double latitude, double longitude) {
+    if (!isValidLatitude(latitude)) {
+      getDispatchDelegate().dispatchErrorOccurredEvent(this, "SetEndLocation",
+          ErrorMessages.ERROR_INVALID_LATITUDE, latitude);
+    } else if (!isValidLongitude(longitude)) {
+      getDispatchDelegate().dispatchErrorOccurredEvent(this, "SetEndLocation",
+          ErrorMessages.ERROR_INVALID_LONGITUDE, longitude);
+    } else {
+      endLocation.setCoords(latitude, longitude);
+    }
+  }
+
+  private Double[][] getCoordinates() {
+    Double[][] coords = new Double[2][2];
+    coords[0][0] = startLocation.getLongitude();
+    coords[0][1] = startLocation.getLatitude();
+    coords[1][0] = endLocation.getLongitude();
+    coords[1][1] = endLocation.getLatitude();
+    return coords;
+  }
+
+  private String getCoordinatesFromGeoJson(String geoJson) {
```

Comment:
Rather than doing all of this parsing here, we should consider refactoring the parse logic in `MapFeatureContainerBase` into `GeoJSONUtil` so that it doesn't have to be repeated. Then you can just call the functionality in `GeoJSONUtil`